### ISSUE TITLE
[HOLD OFFLINE + Report Logic Refactor] Unread messages stays unread on opening the chat when offline - Reported by @thesahindia

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7727]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app as account A 
2. Send few messages from Account B. Do not open messages 
3. Set account A device offline
4. Navigate to new message tread 
5. Go back to LHN 

## Expected Result:
Messages should be marked as read 

## Actual Result:
Messages still show unread 

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platform:

Where is this issue occurring?

- iOS
- Android

**Version Number:** 1.1.38-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/153809298-347e68c5-52ca-4ba4-bbd2-a12284c425a0.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @thesahindia
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1643317442138000

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @JmillsExpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by JmillsExpensify** [comment]
This is a good issue, it also happens to me frequently on intermittent connections. That said, I'm uncertain is this is an external or internal issue, so I'm adding the engineering label for a second opinion.

**Comment by MelvinBot** [comment]
Current assignee @JmillsExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by PauloGasparSv** [comment]
Hi @JmillsExpensify, I discussed the issue with @pecanoro and it looks like an External issue since it is an offline thing. So I've added the external label.

**Comment by MelvinBot** [comment]
@JmillsExpensify Eep! 4 days overdue now. Issues have feelings too...

**Comment by JmillsExpensify** [comment]
Still waiting on volunteers

**Comment by parasharrajat** [comment]
@JmillsExpensify Exported label is missing from this one.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcaaron (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by JmillsExpensify** [comment]
Thank you

**Comment by JmillsExpensify** [comment]
Doubled the rate of this Upwork job to $500.

**Comment by JmillsExpensify** [comment]
Going to post this issue in #expensify-open-source and then if we don't have any volunteers, I will increase the payout for this job.

**Comment by JmillsExpensify** [comment]
Still no takers. Increased the payout of the job to $1000.

**Comment by mallenexpensify** [comment]
Raised price to $2000

**Comment by mateusbra** [comment]
### PROBLEM

The main issue here is that when we are searching for new reports, we just ignore the report to be read because we assume that it's will be overwritten on the future, so its outdated here:
https://github.com/Expensify/App/blob/7919e24c808dea9dedfe4a2d513365cf092cc1e1/src/libs/actions/Report.js#L1240-L1242

Flow example:
1. User 1 sends few messages to user 2.
2. User 2 triggers `fetchAllReports` by backgrounding/foregrounding the app, which will make the ONYXKEY `isLoadingReportData` to be `true` and ignore the reading.
3. When we receive all report data we set all the report data with the values for the moment the function `fetchAllReports` was called.

The problem here is that we are ignoring whatever user 2 does between the moment we call `fetchAllReports` and it returns us the report data.


 ### PROPOSAL
1. We need to remove this:
https://github.com/Expensify/App/blob/7919e24c808dea9dedfe4a2d513365cf092cc1e1/src/libs/actions/Report.js#L1240-L1242

Now we need to know if  user changed `lastReadSequenceNumber` between the time we called `fetchAllReports`.

How do we know that?
If user readed the last messages while loading, we will have a `lastReadSequenceNumbers[report.reportID]` greater then the report's lastReadSequenceNumber provided by API which will be our `lastReadSequenceNumber`.

2. So we should compare if `lastReadSequenceNumbers[report.reportID] > lastReadSequenceNumber` within https://github.com/Expensify/App/blob/7919e24c808dea9dedfe4a2d513365cf092cc1e1/src/libs/actions/Report.js#L97-L101
adding to line 102:
```js
     if(lastReadSequenceNumbers[report.reportID] > lastReadSequenceNumber){
        unreadActionCount = report.reportActionCount - lastReadSequenceNumbers[report.reportID] - ReportActions.getDeletedCommentsCount(report.reportID, lastReadSequenceNumbers[report.reportID]);
        return Math.max(0, unreadActionCount);
    }
```

We will use the same logic as used here, but using lastReadSequenceNumbers[report.reportID], which will be the correctly lastReadSequenceNumber for that report:
https://github.com/Expensify/App/blob/7919e24c808dea9dedfe4a2d513365cf092cc1e1/src/libs/actions/Report.js#L116-L118

**Comment by rushatgabhane** [comment]
on it ☝️

**Comment by marcaaron** [comment]
I feel it needs to be said that this code could use a refactor. Maybe it _should_ be made internal? @mallenexpensify any idea how many issues we have related to "read" status we have out right now? 

There's [this one](https://github.com/Expensify/App/pull/6764) that @mateusbra has been working with @parasharrajat for months and honestly it just feels like if an issue like that is taking 3+ months then maybe we got the design wrong and should go back to the drawing board.

**Comment by rushatgabhane** [comment]
Sounds good. We also have this from slack: [unread bold in LHN is not going away, even if I am in that chat](https://expensify.slack.com/archives/C01GTK53T8Q/p1647467584945749)

**Comment by parasharrajat** [comment]
Yup. I do agree with refactoring the code. I can see that a lot of code is responsible for marking the messages read and different functions are setting new line indicator. It has become quite complex to see the exact flow of information that is used to set the unread count. ReportActionsView has become a mess to manage unread/read behavior. 

On https://github.com/Expensify/App/pull/6764, we never found the proper solution that was expected. There is still a delay before the new marker is set. Due to the PR was stalling for months and in absence of new solutions from the author and me, I decided to close it by choosing the best-found solution. 

I also feel that it would be really helpful to have a high-level design doc for read/unread and unread count functionality.

**Comment by mateusbra** [comment]
It seems like many times we don't consider the flow of when user could be offline or when we are loading data from API, 
As an external contributor I feel it should be redesigned/refactored and agreed with:
> I also feel that it would be really helpful to have a high-level design doc for read/unread and unread count functionality.

**Comment by marcaaron** [comment]
Ok cool. I'm going to suggest that we HOLD this for the Offline project and also look into some refactoring of the Report page (which everyone agrees is a mess - but we'll need to get a bit more specific about what is wrong and how to improve it exactly).

**Comment by JmillsExpensify** [comment]
Still holding on the offline project, though hopefully for just another week or two. 🙏

**Comment by JmillsExpensify** [comment]
Still holding for the offline project.

**Comment by JmillsExpensify** [comment]
Still holding for offline initiatives.

**Comment by JmillsExpensify** [comment]
Still held on offline

**Comment by pecanoro** [comment]
@JmillsExpensify Moving to monthly since it's on hold so you don't have to post every week about it!

**Comment by rushatgabhane** [comment]
@Expensify/contributor-plus feel free to pick this up, I'm unassigning because I'll be OOO.

All proposals here have been triaged so far.

**Comment by mananjadhav** [comment]
@JmillsExpensify @pecanoro  Is this still on hold?

**Comment by JmillsExpensify** [comment]
Yes, still on hold. We're getting really close to unblocking it though. I'd estimate another couple of weeks.

**Comment by JmillsExpensify** [comment]
Still held on offline, and also still a week or more out.

**Comment by JmillsExpensify** [comment]
Still holding on offline, though we're through the internal review process and closer to moving to implementation.

**Comment by michaelhaxhiu** [comment]
@JmillsExpensify I have been working through these lingering Offline Mode issues that were made by contributors (before we actually roadmapped and prioiritized Offline First). 

I've been paying and closing these today, as they are basically just cluttering our job list. We intend to build much of offline first internally, and then if we have aspects where we can outsource I'm sure we'll do that. But I think we should just pay and close this "bug find" at this point for simplicity. Let me now if you agree and I can take it over.

**Comment by JmillsExpensify** [comment]
> But I think we should just pay and close this "bug find" at this point for simplicity. Let me now if you agree and I can take it over.

💯 Love it. This is a great suggestion.

**Comment by michaelhaxhiu** [comment]
@thesahindia can you please apply for this job in Upwork - https://www.upwork.com/jobs/~019fa14854ad47fca7

**Comment by thesahindia** [comment]
@michaelhaxhiu, applied ✔️

**Comment by michaelhaxhiu** [comment]
Done deal, closing.
